Update timestep after postprocessing

class4kayaker opened the pull request:
Adjust the core timestep logic to update the timestep values just before updating the model time, which is after calling the post-processors. This fixes #929.

In order to maintain statistics file structure, I believe that it is necessary to compute the timestep at the current location, so a single new double was required. The logic was slightly reworked to allow it to be const. Given this, updating the timestep was delayed until just before updating the time.

Given that areas that might be affected, there should probably be some examination of whether the locations that are obviously using the new timestep should be.

Looking through the plugins in the actual post-processing folder, it appears that none of them are dependent on the current approach. The particle code has been adjusted for correct behavior.

gassmoeller commented:
I am curious if this passes all tests (there are quite some pitfalls with plugins using time information). Looks good otherwise.

/run-tests

class4kayaker commented:
All of the current issues appear to be tied to the artificial viscosity computation, and most of the results other than artificial_postprocess appear to be small. I could not identify where the timestep might be used directly, but given that it is calling the simulator's artificial viscosity calculation, the new values are likely correct.

At this point, I think it would make the most sense to look over where the new timestep is used to ensure that it should be used there. There are at least a few locations that I believe might currently be incorrect, including possibly the time update.

gassmoeller commented:
I would guess the difference comes from the `assembly.cc::compute_advection_system_residual()` function that uses `get_old_timestep()` which is now different from before. It is not obvious on first sight, which value should be used right here, because it is used both during assembly of the matrices (where it uses the same timestep before and after this PR), and during postprocessing and mesh refinement (where it now uses a different value than before). 
However on closer look I agree, supposedly the new way is correct. For postprocessing we definitely want to output the artificial viscosity that was used to compute this timestep (to be consistent with the solution that we output). For mesh refinement  one can make a similar argument.
Could you take a look and see if I understood this right?

class4kayaker commented on source/simulator/core.cc:
```diff
@@ -2379,19 +2379,18 @@ namespace aspect
+          statistics.add_value("Time step size (years)", new_time_step / year_in_seconds);
         else
-          statistics.add_value("Time step size (seconds)", time_step);
+          statistics.add_value("Time step size (seconds)", new_time_step);
```
Should we be listing the current timestep for each iteration's statistics data instead of the next timestep?

gassmoeller replied:
I do not think there is a *best* solution for this, but since so far we listed the 'next' time step size (i.e. for time step n output the time step size from n to n+1, and not from n-1 to n) I would say we keep the behavior. So your change is correct.

bangerth replied:
Let's leave it like this for now, but it would be nice if in a follow-up patch we could change this to ensure that the time steps we list in the statistics file add up to the current time. I'll open an issue for this.

class4kayaker commented on source/simulator/core.cc:
```diff
@@ -2422,11 +2421,11 @@ namespace aspect
             while ((parameters.additional_refinement_times.size() > 0)
                    &&
-                   (parameters.additional_refinement_times.front () < time+time_step))
+                   (parameters.additional_refinement_times.front () < time+new_time_step))
```
Should we be incrementing the time by the new timestep? If this is incorrect, then the time update is also incorrect.

gassmoeller replied:
Yes we should, your change is correct as well. The reason behind this is that we have already solved the timestep above, and here we are checking, whether the *next* timestep will be after an additional refinement time (everything that is happening here is preparing the variables for the next timestep).

class4kayaker commented:
I concur with your logic and have confirmed that this is now what is done for the artificial viscosity.

Looking through the places I changed to make use of the new timestep, I noticed that there are a number of locations where we may actually want to use time_step rather than new_time_step. I have added comments for these, and would appreciate external confirmation of whether I am overlooking a convention.

gassmoeller commented:
I think your changes are all correct. We could consider changing `new_timestep` to `next_timestep` to make the meaning more clear, but I am not sure if that is worth the effort. Tester looks good so far. Lets wait a day for others to comment if they want, otherwise I think this is good to go.

gassmoeller commented:
@tjhei: Is your tester all-right? aspect-8.4.1b  passed all gcc tests, but 8.4.1 had a lot of timeouts.

class4kayaker commented:
Thank you for confirming that there were not any issues with the locations I actually changed. I knew I was matching the past behavior, but I wanted an external confirmation that the adjustment was not making incorrect behavior the obvious intent.

gassmoeller commented:
@class4kayaker: I think the tester had a problem, but seems fixed now. If you have the time, please squash all commits into one, and push the branch. That should trigger a new test run.

gassmoeller commented:
Thanks for cleaning this up! Tester is happy and the changed test results seem reasonable, so I am going to merge tomorrow morning, if nobody objects.

bangerth commented:
Since @gassmoeller was going to merge and since I'm ok with the change as well, I'm just going to do it for him.
